heat_pump_adoption_modelling/pipeline/MCS/mcs_epc_joining.py

```diff
@@ -0,0 +1,322 @@
+    to underscores so that the numeric tokens in e.g.
+    "Flat 3/2" and "Flat 3-2" are treated as a whole later.
+
+    Parameters
+    ----------
+    address : str
+        Address to format.
+
+    Return
+    ----------
+    address : str
+        Formatted address."""
+    if (address is pd.NA) | (address is np.nan):
+        return ""
+    else:
+        # replace / and - with _
+        address = address.replace("/", "_").replace("-", "_")
+        # remove all punctuation other than _
+        address = address.translate(
+            str.maketrans("", "", string.punctuation.replace("_", ""))
+        )
+        return address
+
+
+def extract_numeric_tokens(df):
+    """Extract character strings containing numbers
+    e.g. "45", "3a", "4_1".
+
+    Parameters
+    ----------
+    df : pandas.Dataframe
+        Dataframe with standardised_address field.
+
+    Return
+    ----------
+    token_list : list
+        List of sets of numeric tokens within each standardised_address."""
+
+    token_list = [
+        set(re.findall("\w*\d\w*", address)) for address in df["standardised_address"]
+    ]
+
+    return token_list
+
+
+# this isn't perfect - e.g. "Flat 3, 3 Street Name" would
+# just become {3} - could use a multiset instead
+
+# wonder if single-letter tokens should be in here too
+# for e.g. "Flat A" or whether this would give too many
+# false positives
+
+# ---------------------------------------------------------------------------------
+
+#### PREPROCESSING
+
+
+def prepare_dhps(dhps):
+    """Prepares Dataframe of domestic HP installations by adding
+    standardised_postcode, standardised_address and numeric_tokens fields.
+
+    Parameters
+    ----------
+    dhps : pandas.Dataframe
+        Dataframe with postcode, address_1 and address_2 fields.
+
+    Return
+    ----------
+    dhps : pandas.Dataframe
+        Dataframe containing domestic HP records with added fields."""
+
+    dhps["standardised_postcode"] = [
+        pc.upper().strip() for pc in dhps["postcode"].fillna("unknown")
+    ]
+
+    dhps["standardised_address"] = [
+        rm_punct(add1).lower().strip() + " " + rm_punct(add2).lower().strip()
+        for add1, add2 in zip(dhps.address_1.fillna(""), dhps.address_2.fillna(""))
+    ]
+
+    dhps["numeric_tokens"] = extract_numeric_tokens(dhps)
+
+    return dhps
+
+
+def prepare_epcs(epcs):
+    """Prepares Dataframe of EPC records by adding
+    standardised_postcode, standardised_address and numeric_tokens fields.
+
+    Parameters
+    ----------
+    epcs : pandas.Dataframe
+        Dataframe with POSTCODE and ADDRESS1 fields.
+
+    Return
+    ----------
+    epcs : pandas.Dataframe
+        Dataframe containing EPC records with added fields."""
+
+    # Remove spaces, uppercase and strip whitespace from
+    # postcodes in order to exact match on this field
+    epcs["standardised_postcode"] = [
+        postcode.replace(" ", "") for postcode in epcs["POSTCODE"].fillna("UNKNOWN")
+    ]
```

Comment:
I don't see the uppercase happening...?

```suggestion
    epcs["standardised_postcode"] = [
        postcode.upper().replace(" ", "") for postcode in epcs["POSTCODE"].fillna("UNKNOWN")
    ]
```

Reply:
Good point! EPC postcodes tend to be uppercase anyway but will add just in case